File: src/main/java/com/hiberus/anaya/redmineapi/UrlJSON.java
package com.hiberus.anaya.redmineapi;

import org.json.JSONObject;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

/**
 * Utilities for urls with JSON
 */
public class UrlJSON {

    /**
     * Reads JSON data from an url
     * From https://stackoverflow.com/a/4308662
     *
     * @param url url to load
     * @return the json data returned
     * @throws IOException on network errors
     */
    public static JSONObject get(String url) throws IOException {
        try (InputStream is = new URL(url).openStream()) {
            return new JSONObject(readAll(new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8))));
        }
    }

    /**
     * Performs a PUT to an url with JSON data
     *
     * @param url    url to put to
     * @param object the json data to put
     * @return the response status code
     * @throws IOException on network errors
     */
    public static int put(String url, JSONObject object) throws IOException {
        return send(url, "PUT", object);
    }


    /**
     * Performs a POST to an url with JSON data
     *
     * @param url    url to post to
     * @param object the json data to post
     * @return the response status code
     * @throws IOException on network errors
     */
    public static int post(String url, JSONObject object) throws IOException {
        return send(url, "POST", object);
    }


    /**
     * Performs a DELETE to an url
     *
     * @param url url to delete
     * @return the response status code
     * @throws IOException on network errors
     */
    public static int delete(String url) throws IOException {
        return send(url, "DELETE", null);
    }

    /* ------------------------- utils ------------------------- */

    private static int send(String url, String method, JSONObject body) throws IOException {
        // performs a send to an url

        // prepare connection
        HttpURLConnection http = (HttpURLConnection) new URL(url).openConnection();
        http.setRequestMethod(method);
        if (body != null) {
            // append body
            http.setDoOutput(true);
            http.setRequestProperty("Content-Type", "application/json");
            http.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        }

        // get
        int responseCode = http.getResponseCode();
        System.out.println(responseCode + ": " + http.getResponseMessage());

        // end
        http.disconnect();
        return responseCode;
    }

    private static String readAll(Reader rd) throws IOException {
        // converts a reader to a string
        // from https://www.baeldung.com/java-convert-reader-to-string
        char[] arr = new char[8 * 1024];
        StringBuilder buffer = new StringBuilder();
        int numCharsRead;
        while ((numCharsRead = rd.read(arr, 0, arr.length)) != -1) {
            buffer.append(arr, 0, numCharsRead);
        }
        return buffer.toString();
    }
}

File: src/main/java/com/hiberus/anaya/redmineapi/UrlJSON.kt
package com.hiberus.anaya.redmineapi

import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

/* ------------------------- Utilities for urls with JSON ------------------------- */

/**
 * Reads JSON data from this url
 *
 * @return the json data returned
 * @throws IOException on network errors or invalid result (not 200-OK)
 */
@Throws(IOException::class) // TODO: replace all of this with async/await?
fun URL.getJSON() = (openConnection() as HttpURLConnection).run {
    if (responseCode != 200) throw IOException("Returned $responseCode: $responseMessage")
    else openStream().use {
        JSONObject(it.bufferedReader().readText())
    }
}

/**
 * POSTs this JSON data to that url
 *
 * @param url    url to post to
 * @return the response status code
 * @throws IOException on network errors
 */
@Throws(IOException::class)
fun JSONObject.postTo(url: URL) = url.send("POST", this)

/**
 * PUTs this JSON data to an url
 *
 * @param url    url to put to
 * @return the response status code
 * @throws IOException on network errors
 */
@Throws(IOException::class)
fun JSONObject.putTo(url: URL) = url.send("PUT", this)

/**
 * Performs a DELETE to this url
 *
 * @return the response status code
 * @throws IOException on network errors
 */
@Throws(IOException::class)
fun URL.delete(): Int = send("DELETE", null)

/* ------------------------- private ------------------------- */

/**
 * performs a 'send' to an url
 */
@Throws(IOException::class)
private fun URL.send(method: String, body: JSONObject?) =
    (openConnection() as HttpURLConnection).run {
        // prepare connection
        requestMethod = method
        body?.let {
            // append body
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            outputStream.write(it.toString().toByteArray(StandardCharsets.UTF_8))
        }

        // connect (call not needed, but easier to follow code flow)
        connect()

        // perform debug
        println("$responseCode: $responseMessage")

        // end
        disconnect()

        // return responseCode
        responseCode
    }

/* ------------------------- utils ------------------------- */

/**
 * Compare value with another one, if different execute function
 */
inline fun <T> T.ifNot(compareTo: T, execute: () -> Unit) {
    if (this != compareTo) execute()
}

/**
 * Map a JSONArray as a list of JSONObject
 */
internal fun JSONArray.mapAsObjects() = List(length()) { i -> getJSONObject(i) }